core/src/main/java/jenkins/util/VirtualFile.java

```diff
@@ -617,13 +685,40 @@ public static VirtualFile forFilePath(final FilePath f) {
                     throw new IOException(x);
                 }
             }
-            @Override public Collection<String> list(String includes, String excludes, boolean useDefaultExcludes) throws IOException {
+
+            @Override public boolean supportsQuickRecursiveListing() {
+                return this.f.getChannel() == FilePath.localChannel;
+            }
+
+            @Override public @Nonnull List<VirtualFile> listOnlyDescendants() throws IOException {
                 try {
-                    return f.act(new Scanner(includes, excludes, useDefaultExcludes));
+                    if (!isDescendant("")) {
+                        return Collections.emptyList();
+                    }
+
+                    List<FilePath> children = f.list();
+                    List<VirtualFile> legalChildren = new ArrayList<>(children.size());
+                    for (FilePath child : children){
+                        if (isDescendant(child.getName())) {
+                            FilePathVF legalChild = new FilePathVF(child, this.root);
+                            legalChild.cacheDescendant = true;
+                            legalChildren.add(legalChild);
+                        }
+                    }
+
+                    return legalChildren;
                 } catch (InterruptedException x) {
                     throw new IOException(x);
                 }
             }
+
+        @Override public Collection<String> list(String includes, String excludes, boolean useDefaultExcludes) throws IOException {
+            try {
+                return f.act(new Scanner(includes, excludes, useDefaultExcludes));
```

Comment:
Did you mean:

```suggestion
                return f.act(new Scanner2(includes, excludes, useDefaultExcludes));
```

Reply:
No it was just during my testing. I changed the behavior of the Scanner that now consider only the descendants of the root. I thought initially I will need to split the behavior in two but that was not the case in "final" version.